src/Nethermind/Nethermind.BeaconNode/AttestationProducer.cs

```diff
@@ -0,0 +1,126 @@
+
+            if (validatorIndex == null)
+            {
+                throw new Exception($"Can not find validator index for public key {validatorPublicKey}");
+            }
+
+            // TODO: May need a more efficient way to try and find the committee and position within the committee.
+            // Some of this may already be cached in Validator Assignments (generally stable for an epoch),
+            // but not with the index within the committee. Easy enough to extend and use the same cache.
+
+            IReadOnlyList<ValidatorIndex> committee =
+                _beaconStateAccessor.GetBeaconCommittee(headState, headState.Slot, index);
+            int committeeSize = committee.Count;
+
+            int? committeeMemberIndexOfValidator = null;
+            for (int committeeMemberIndex = 0; committeeMemberIndex < committee.Count; committeeMemberIndex++)
+            {
+                if (committee[committeeMemberIndex] == validatorIndex)
+                {
+                    committeeMemberIndexOfValidator = committeeMemberIndex;
+                    break;
+                }
+            }
+
+            if (committeeMemberIndexOfValidator == null)
+            {
+                throw new Exception($"Validator index {validatorIndex} is not a member of committee {index}");
+            }
+
+            Root beaconBlockRoot = _cryptographyService.HashTreeRoot(headBlock);
+
+            Checkpoint source = headState.CurrentJustifiedCheckpoint;
+
+            Epoch currentEpoch = _beaconStateAccessor.GetCurrentEpoch(headState);
+            Slot startSlot = _beaconChainUtility.ComputeStartSlotOfEpoch(currentEpoch);
+            Root epochBoundaryBlockRoot;
+            if (startSlot == headState.Slot)
+            {
+                epochBoundaryBlockRoot = beaconBlockRoot;
+            }
+            else
+            {
+                epochBoundaryBlockRoot = _beaconStateAccessor.GetBlockRootAtSlot(headState, startSlot);
+            }
+
+            Checkpoint target = new Checkpoint(currentEpoch, epochBoundaryBlockRoot);
+
+            AttestationData attestationData = new AttestationData(slot, index, beaconBlockRoot, source, target);
+
+            var aggregationBits = new BitArray(committeeSize);
+            aggregationBits.Set(committeeMemberIndexOfValidator.Value, true);
+
+            var attestation = new Attestation(aggregationBits, attestationData, BlsSignature.Zero);
+
+            return attestation;
+        }
+
+        // De-duplicate from ValidatorAssignments
+        private ValidatorIndex? FindValidatorIndexByPublicKey(BeaconState state, BlsPublicKey validatorPublicKey)
+        {
+            for (int index = 0; index < state.Validators.Count; index++)
+            {
+                if (state.Validators[index].PublicKey.Equals(validatorPublicKey))
+                {
+                    return new ValidatorIndex((ulong) index);
+                }
+            }
```

Comment:
dunno if we can use linq but if we do 

```
if(state.Validators.Any(validator => validator.PublicKey.Equals([...])))
     return ...
```

Reply:
your suggested change introduces a closure and unnecessary memory allocation - we use for loops to keep memory in check

Reply:
true, didn't think of that this way :X

Reply:
This is something that would benefit from caching, not yet implemented/optimised. It is mostly a naive algorithm (probably from the spec).

In practice the number of validators will be large, hundreds of thousands, but the range handled by a particular beacon node may be smaller, maybe just hundreds.

Reply:
Hmmm... once assigned, an index never changes, so for a given state and all descendants it could be cached... however if there was a reorg and switch to another state the index could be different. (Index effectively comes from Eth1 deposit, and any Eth1 split with different data could lead to Eth2 split with different data).

The calculation of "is this a child" may be more complicated then simply looping through the list.

If you are caching per-state, or even per epoch (can only change on epoch boundaries), then there may only be ever 2 lookups (one to get duties, one to create the attestation; proposal is very rare), so you are weighing 2 lookups vs 1 lookup + small cache vs large cache.

More complicated than a simple change. Needs benchmarking to optimise.

Reply:
Maybe the getduties loop, which usually operates on a collection so can loop through once, could cache the relevant lookups for the epoch. We already calculate the epoch root below.

In fact, it is probably the same calc as get duties, as the duties already has most of the detail (committee index, etc), just need the member index.

Reply:
Added as a card to the project board.